packages/plugin-git/sources/gitUtils.ts

```diff
@@ -317,3 +318,76 @@ export async function clone(url: string, configuration: Configuration) {
     return directory;
   });
 }
+
+export async function fetchRoot(initialCwd: PortablePath) {
+  // Note: We can't just use `git rev-parse --show-toplevel`, because on Windows
+  // it may return long paths even when the cwd uses short paths, and we have no
+  // way to detect it from Node (not even realpath).
+
+  let match: PortablePath | null = null;
+
+  let cwd: PortablePath;
+  let nextCwd = initialCwd;
+  do {
+    cwd = nextCwd;
+    if (await xfs.existsPromise(ppath.join(cwd, `.git` as Filename)))
+      match = cwd;
+    nextCwd = ppath.dirname(cwd);
+  } while (match === null && nextCwd !== cwd);
+
+  return match;
+}
+
+export async function fetchBase(root: PortablePath, {baseRefs}: {baseRefs: Array<string>}) {
+  if (baseRefs.length === 0)
+    throw new UsageError(`Can't run this command with zero base refs specified.`);
+
+  const ancestorBases = [];
+
+  for (const candidate of baseRefs) {
+    const {code} = await execUtils.execvp(`git`, [`merge-base`, candidate, `HEAD`], {cwd: root});
+    if (code === 0) {
+      ancestorBases.push(candidate);
+    }
+  }
+
+  if (ancestorBases.length === 0)
+    throw new UsageError(`No ancestor could be found between any of HEAD and ${baseRefs.join(`, `)}`);
+
+  const {stdout: mergeBaseStdout} = await execUtils.execvp(`git`, [`merge-base`, `HEAD`, ...ancestorBases], {cwd: root, strict: true});
+  const hash = mergeBaseStdout.trim();
+
+  const {stdout: showStdout} = await execUtils.execvp(`git`, [`show`, `--quiet`, `--pretty=format:%s`, hash], {cwd: root, strict: true});
+  const title = showStdout.trim();
+
+  return {hash, title};
+}
+
+// Note: This returns all changed files from the git diff, which can include
+// files not belonging to a workspace
+export async function fetchChangedFiles(root: PortablePath, {base, project}: {base: string, project: Project}) {
+  const {stdout: localStdout} = await execUtils.execvp(`git`, [`diff`, `--name-only`, `${base}`], {cwd: root, strict: true});
+  const trackedFiles = localStdout.split(/\r\n|\r|\n/).filter(file => file.length > 0).map(file => ppath.resolve(root, npath.toPortablePath(file)));
+
+  const {stdout: untrackedStdout} = await execUtils.execvp(`git`, [`ls-files`, `--others`, `--exclude-standard`], {cwd: root, strict: true});
+  const untrackedFiles = untrackedStdout.split(/\r\n|\r|\n/).filter(file => file.length > 0).map(file => ppath.resolve(root, npath.toPortablePath(file)));
+
+  return [...new Set([...trackedFiles, ...untrackedFiles].sort())];
+}
+
+// Note: .pnp.cjs and lockfile are excluded from workspace change detection
+// as they can be modified by changes to any workspace manifest file.
+export async function fetchChangedWorkspaces(root: PortablePath, {base, project}: {base: string, project: Project}) {
+  const ignoredFiles = [`.pnp.cjs`, project.configuration.get(`lockfileFilename`)];
```

Comment:
Is it okay to hardcode `.pnp.cjs`? I couldn't find any config options or constants to use instead.

Reply:
There's a hook, [`populateYarnPaths`](https://github.com/yarnpkg/berry/blob/5411f76bcd89d1d6f430f4bd0e16146ce9fdd370/packages/yarnpkg-core/sources/Plugin.ts#L131-L137), which lets you get all the Yarn artifacts. See here:

https://github.com/yarnpkg/berry/blob/7ae458b8165ad53a8ef9db0060cbb6de73305768/packages/plugin-pack/sources/packUtils.ts#L183-L195

Reply:
Also note that the lockfile path needs to be joined to `project.cwd` (like in the snippet I shared).

Reply:
Ah great. Thanks for the pointers. I added the paths using that hook and updated the lockfile path to be joined to `project.cwd`.
Fixed in 767fa1270516be7f819d29cbd23b1fbab0286676.

Reply:
I didn't bother adding some of the other paths like `installStatePath` as I figured that is advised to be gitignored anyway.
Are there any other paths you would recommend adding here too?
I'm thinking maybe `cacheFolder` and `yarnPath`.

Reply:
I updated this to include more yarn artifact paths. I intentionally excluded the rc file and yarn path as those are likely not modified by workspaces, and changes to the package manager version might warrant re-running scripts in all workspaces.